File: spring-boot-oauth2-react-client/spring-boot-oauth2-react-client-frontend-api/src/main/java/com/onlyteo/sandbox/cache/ReferrerAwareHttpSessionRequestCache.java
package com.onlyteo.sandbox.cache;

import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import org.springframework.http.HttpHeaders;
import org.springframework.security.web.savedrequest.HttpSessionRequestCache;
import org.springframework.security.web.savedrequest.SimpleSavedRequest;
import org.springframework.util.StringUtils;

/**
 * Custom request cache implementation that will save the URL from the {@code Referer} HTTP header. If the
 * {@code Referer} header is not present in the request it will save the original request URL.
 */
public class ReferrerAwareHttpSessionRequestCache extends HttpSessionRequestCache {

    private static final String SAVED_REQUEST_ATTRIBUTE_NAME = "SPRING_SECURITY_SAVED_REQUEST";

    public ReferrerAwareHttpSessionRequestCache() {
        setSessionAttrName(SAVED_REQUEST_ATTRIBUTE_NAME);
    }

    @Override
    public void saveRequest(final HttpServletRequest request, final HttpServletResponse response) {
        final var referrer = request.getHeader(HttpHeaders.REFERER);
        final var redirectUrl = StringUtils.hasText(referrer) ? referrer : request.getRequestURL().toString();
        request.getSession().setAttribute(SAVED_REQUEST_ATTRIBUTE_NAME, new SimpleSavedRequest(redirectUrl));
    }
}

File: spring-boot-oauth2-react-client/spring-boot-oauth2-react-client-frontend-api/src/main/kotlin/com/onlyteo/sandbox/cache/ReferrerAwareHttpSessionRequestCache.kt
package com.onlyteo.sandbox.cache

import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import org.springframework.http.HttpHeaders
import org.springframework.security.web.savedrequest.HttpSessionRequestCache
import org.springframework.security.web.savedrequest.SimpleSavedRequest
import org.springframework.util.StringUtils

/**
 * Custom request cache implementation that will save the URL from the {@code Referer} HTTP header. If the
 * {@code Referer} header is not present in the request it will save the original request URL.
 */
class ReferrerAwareHttpSessionRequestCache : HttpSessionRequestCache() {

    private val sessionAttributeName = "SPRING_SECURITY_SAVED_REQUEST"

    init {
        setSessionAttrName(sessionAttributeName)
    }

    override fun saveRequest(request: HttpServletRequest, response: HttpServletResponse) {
        val referrer = request.getHeader(HttpHeaders.REFERER)
        val redirectUrl = if (StringUtils.hasText(referrer)) referrer else request.requestURL.toString()
        request.session.setAttribute(sessionAttributeName, SimpleSavedRequest(redirectUrl))
    }
}